gate-oauth2/src/main/java/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.java

```diff
@@ -0,0 +1,304 @@
+    if (email == null || !permissionService.isEnabled()) {
+      return false;
+    }
+    try {
+      return Optional.ofNullable(Retrofit2SyncCall.execute(front50Service.getServiceAccounts()))
+          .orElse(new ArrayList<>())
+          .stream()
+          .anyMatch(sa -> email.equalsIgnoreCase(sa.getName()));
+    } catch (SpinnakerServerException e) {
+      log.warn("Could not get list of service accounts.", e);
+    }
+    return false;
+  }
+
+  private static boolean valueMatchesConstraint(Object value, String requiredVal) {
+    if (value == null) {
+      return false;
+    }
+
+    if (isRegexExpression(requiredVal)) {
+      return String.valueOf(value).matches(mutateRegexPattern(requiredVal));
+    }
+
+    return value.equals(requiredVal);
+  }
+
+  public boolean hasAllUserInfoRequirements(Map<String, Object> details) {
+    if (userInfoRequirements == null || userInfoRequirements.isEmpty()) {
+      return true;
+    }
+
+    Map<String, String> invalidFields =
+        userInfoRequirements.entrySet().stream()
+            .filter(
+                entry -> {
+                  String reqKey = entry.getKey();
+                  String reqVal = entry.getValue();
+                  Object value = details.get(reqKey);
+
+                  if (value instanceof Collection<?> collection) {
+                    return collection.stream()
+                        .noneMatch(item -> valueMatchesConstraint(item, reqVal));
+                  }
+                  return !valueMatchesConstraint(value, reqVal);
+                })
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    if (!invalidFields.isEmpty() && log.isDebugEnabled()) {
+      log.debug(
+          "Invalid userInfo response: "
+              + invalidFields.entrySet().stream()
+                  .map(
+                      entry ->
+                          "got "
+                              + entry.getKey()
+                              + "="
+                              + details.get(entry.getKey())
+                              + ", wanted "
+                              + entry.getValue())
+                  .collect(Collectors.joining(", ")));
+    }
+
+    return invalidFields.isEmpty();
+  }
+
+  public static boolean isRegexExpression(String val) {
+    if (val.startsWith("/") && val.endsWith("/")) {
+      try {
+        Pattern.compile(val);
+        return true;
+      } catch (PatternSyntaxException ignored) {
+        return false;
+      }
+    }
+
+    return false;
+  }
+
+  public static String mutateRegexPattern(String val) {
+    // "/expr/" -> "expr"
+    return val.substring(1, val.length() - 1);
+  }
+
+  protected List<String> getRoles(Map<String, Object> details) {
+    if (userInfoMapping == null || userInfoMapping.getRoles() == null) {
+      return List.of();
+    }
+
+    Object roles = details.getOrDefault(userInfoMapping.getRoles(), List.of());
```

Comment:
I think this is slightly different from the groovy:
```
def roles = details[userInfoMapping.roles] ?: []
```
From my understanding, if the value of the `userInfoMapping.roles` key is null, the groovy code assigns the empty list to roles, but the java code assigns null.

Reply:
updated the code.

Reply:
https://github.com/spinnaker/gate/pull/1871/commits/da85da01531e6c40f7a28a062076ee490b851cd4